### DISCUSSION TITLE
Covariate consideration for the estimation process

### DISCUSSION THREAD

**Discussion by mcruf** [inlabru-org/inlabru#182] [Q&A]
Hi All,

I'm currently building a LGCP model within inlabru, but I'm a bit unsure on how it works internally with respect to covariates extraction. If I understand it correctly, I'm afraid it might not work for my current case study, and hence any insights on how I could tailor it to my needs would be very much appreciated.

There are essentially two distinct processes on how covariate values can be considered:
1) Estimation - covariate values have to be available for the observed data points.
2) Prediction - covariate values have to be available for the entire prediction grid.

As far as I understood from some existing tutorials (eg., [Gorillas dataset example](https://inlabru-org.github.io/inlabru/articles/web/2d_lgcp_covars.html)), inlabru does both above-mentioned processes automatically when one supplies a covariate file (e.g., a raster file as a _SpatialPixelsDataFrame_ format).

For the estimation process, this is all fine as long as one has information that comes from a point. However, how about information that comes from other geometries, such as lines and polygons? 

In my case study, for instance, I have information that is provided at three levels: point, line, and polygon. A simplistic way to deal with this would be to to take the centroid of the line/polygon and use that as a reference to extract the covariate. The problem, however, is that it does not capture the full variability of the covariate. My approach so far has been to take median covariate values for polygons. Lines, in turn, are discretized into multiple points (e.g., 10 m bins), and median values calculated for the points belonging to a given line. Ultimately, I take these values and allocate them to the centroid of these geometries. 

Of course the latter aspect is not very precise, and I should likely build a multi-likelihood LGCP, one for each geometry type. 
Irrespective of that, the problem of covariate extraction for the estimation process remains. 

Within INLA, this is fairly easy to specify with the inla.stacks function (estimation & prediction stacks). However, I don't see how I could do this within inlabru.

Any ideas? 

Many thanks in advance for your help and time!

-Marie

**Comment by finnlindgren** [comment]
For LGCPs, it's not quite like that for estimation. For estimation, the covariate values need to be available _throughout the observed domain_, i.e. anywhere where there _could_ have been points observed. For prediction, the covariates only need to be available wherever you want to do prediction, which could be either a larger or smaller region.

The normal way of dealing with this is to store the covariate as some kind of raster (the automatic handling currently supports `SpatialGrid/PixelsDataFrame` data and `SpatRaster` data (from the `terra` package), and the corresponding `eval_spatial` methods are called either internally or explicitly in the component input.

If you have a covariate that should be treated as piecewise constant, over polygons, then the cleanest solution would be to add a method for e.g. `sf` objects of polygons; we discussed this internally recently, for this use case, but haven't had time to implement it yet.

But what you describe for lines and polygons seems to be a bit different, in that your covariate is available on a _finer_ scale? In that case, this is part of ongoing research into how to best implement the likelihood integration; one approach is to smooth the covariate onto the spde mesh (or strictly; onto the integration points), but this hasn't been automated, and the resulting methods also vary depending on what covariate values are used at the detected points.

It also seems that you're describing _aggregated count_ data, and not _point pattern_ data; if you don't have or use the actual point locations, but only information aggregated to lines or polygons, that's not an (LG)CP, but rather a plain count data model.  The two types of observation models are of course closely related, but their practical implementation differs.  We are currently expanding the integration code in inlabru to more directly support these types of observations.

With inlabru, we normally avoid prediction stacks, and instead use posterior sampling with `generate()` and `predict()`, as that is more general and easier to automate.

**Reply by finnlindgren** [reply]
I've added a `eval_spatial` method for `sf` data for point-in-polygon data lookup (this may not be what you were looking for, but it covers some of the use cases I mentioned), in commit a496db9

**Reply by mcruf** [reply]
Hi Finn,

Thank you for your prompt and very insightful reply.

I am not entirely sure I understand what you mean by a piecewise constant covariate. For the polygon case, would this mean that the covariate would be treated as a constant value within the polygon domain (even if we know it might not be constant for some cases)? 

I have also looked into your eval_spatial function (thank you so much for that!). However, I'm afraid it won't quite solve my problem at this stage, although it will certainly be very useful for future applications I might work on. Btw, I tested your `eval_spatial` test example, but I get the following error message when  running the code line `nests$something_num <- eval_spatial(plots, nests, layer = "something_num")`:

> Error in UseMethod("eval_spatial") : 
  no applicable method for 'eval_spatial' applied to an object of class "c('sf', 'data.frame')"

I have updated the latest `inlabru`, `sf` and `terra` versions, so not sure why this error is happening.


Also, some more clarification about my case study:

I'm currently dealing with crisis response to flooding events, where information is retrieved from Twitter posts. Twitter information is provided at different spatial granularity, depending on whether the post was geo-tagged or flooding location mentioned explicitly in the tweet text. Geo-tagged posts are the most precise information, as the exact location of flooding is known - so this would be my 'point' geometry-like data. For locations mentioned in the tweet text, information is less precise as it can mention roads (in this case, that would be 'lines') or broader geographical spaces such as neighbourhoods (in this case, that would be the 'polygons') - so the flooding event happened somewhere along the given geometry. Irrespective of the three geometries, the data is aggregated in a sense that it representes the number of tweets for the given location. 

Initially I started to build a geostatistical model testing either Poisson or NB likelihoods, but later on I was advised to try a LGCP model given that the data is essentially 'presence' data (and not presence-absence). This has been therefore my latest attempt for my case study, though I feel there are many challenging factors to account for (e.g., the multiple geometries). I genuinely think that the problem could be treated as a marked point pattern, where the marks could be the number of tweets (aka, 'abundance'). In addition, it seems that a multi-likelihood approach could be more suitable than 1) building a single model irrespective of geometry type, and 2) building separate models for each geometry type. Yet, again, in either of these cases, I face the challenge of the covariate extraction - hence my first question, though the discussion might now lead towards a different direction.

Apologies for the long text, but I thought that providing more details about my case study could help you to understand better my concerns. 

Again, any ideas will be more than welcome and highly appreciated.

-Marie

**Comment by finnlindgren** [comment]
You need the github version of inlabru (currently version 2.7.0.9004) for eval_spatial for sf data. Can also be installed via r-universe. See instructions at https://inlabru-org.github.io/inlabru/

For your data problem, the geolocated data could be modelled via family="cp" (the "LGCP" observation model), yes. And for the others, you need an aggregated count model, like your describe. These can then be combined with `bru(..., like(...), like(...))`, so the issue is mostly how to setup the aggregated count models.

At the moment, you can do what you described for inla, by manually constructing an approximate model by splitting lines etc into smaller pieces, and evaluating the covariates locally, by manually constructing the count model information, but we're working on more automated ways of doing this, that would also properly take into account that for a Poisson count model, one should integrate the exponential of the predictor expression.  For some further information, see the discussion at https://github.com/inlabru-org/inlabru/issues/125
Note that this is partly an ongoing research project.

For covariate extraction, if you have your covariate stored as a terra SpatRaster or a SpatialGridDataGrame, you can use eval_spatial, like `eval_spatial(mycovariate, where = where_to_evaluate)`. The details of this depend on the details of how you setup the model, so I'd need more specific details.

**Reply by mcruf** [reply]
Hi Finn,

Thanks for all the feedback. 
The `eval_spatial` function is working now fine with the github version.

I'll check all the other aspects in more details and refer back to you as soon as I have something more concrete.
